DOCS-2434 Getting Started Continuous Profiler (#11574)
DataDog/documentation: content/en/getting_started/profiler/_index.md

      text: 'Introduction to Datadog'
---

Profiling can make your services faster, cheaper, and more reliable, but if you haven't used a profiler, it can be confusing.

> The guide explains profiling, provides a sample service with a performance problem, and uses the Datadog Continuous Profiler to understand and fix the problem.

Reviewer: ```suggestion
This guide explains profiling, provides a sample service with a performance problem, and uses the Datadog Continuous Profiler to understand and fix the problem.
```

## Overview

A profiler shows how much "work" each function is doing by collecting data about the program as it's running. For example, if infrastructure monitoring shows your app servers are using 80% of their CPU, you may not know why. Profiling shows a breakdown of the work, for example:

| Function      | CPU usage |
|---------------|-----------|
| `doSomeWork`  | 48%       |
| `renderGraph` | 19%       |
| Other         | 13%       |

When working on performance problems, this information is important because many programs spend a lot of time in a few places, which may not not be obvious. Guessing at which parts of a program to optimize causes engineers to spend a lot of time with little results. By using a profiler, you can find exactly which parts of the code to optimize.

If you've used an APM tool, you might think of profiling like a "deeper" tracer that provides a very fine grained view of your code without needing any instrumentation.

> The Datadog Continuous Profiler can track various types of "work", including CPU usage, amount, types of objects being allocated in memory, time spent waiting to acquire locks, amount of network or file I/O, and more. The profile types available depend on the language being profiled.

Reviewer: The and here was correct
```suggestion
The Datadog Continuous Profiler can track various types of "work", including CPU usage, amount and types of objects being allocated in memory, time spent waiting to acquire locks, amount of network or file I/O, and more. The profile types available depend on the language being profiled.
```

When you open it, the visualization of the profile looks similar to this:

{{< img src="tracing/profiling/intro_to_profiling/flame_graph.png" alt="Flame graph">}}

This is a flame graph. The most important things it shows are how much CPU each method used (since this is a CPU profile) and how each method was called. For example, reading from the second row from the top, you see that `Thread.run()` called `QueuedThreadPool$2.run()` (amongst other things), which called `QueuedThreadPool.runjob(Runnable)`, which called `ReservedTheadExecutor$ReservedThread.run()`, and so on.

Zooming in to one area on the bottom of the flame graph, a tooltip shows that roughly 390ms (0.90%) of CPU time was spent within this `parse()` function:

{{< img src="tracing/profiling/intro_to_profiling/flame_graph_parse.png" alt="Flame graph parse() frame">}}

`String.length()` is directly below the `parse()` function, which means that `parse()` calls it. Hover over `String.length()`, to see it took about 112ms of CPU time:

{{< img src="tracing/profiling/intro_to_profiling/flame_graph_length.png" alt="Flame graph String.length() frame">}}

> That means 278ms seconds were spent directly in `parse()`: 390ms - 112ms. That's visually represented by the part of the `parse()` box that doesn't have anything below it.

Reviewer: ```suggestion
That means 278 milliseconds were spent directly in `parse()`: 390ms - 112ms. That's visually represented by the part of the `parse()` box that doesn't have anything below it.
```

It's worth calling out that the flame graph _does not_ represent the progression of time. Looking at this part of the profile, `Gson$1.write()` didn't run before `TypeAdapters$16.write()` but it may not have run after it either.

{{< img src="tracing/profiling/intro_to_profiling/flame_graph_write.png" alt="Flame graph section with write() frames next to each other">}}

 They could have been running concurrently, or the program could have run several calls of one, then several calls of the other, and kept switching back and forth. The flame graph merges together all the times that a program was running the same series of functions so you can tell at a glance which parts of the code were using the most CPU without tons of tiny boxes showing each time a function was called.

> Zoom back out to see that about 87% of CPU usage was within the `replyJSON()` method. Below that, the graphs shows `replyJSON()` and the methods it calls eventually branch into four main code paths ("stack traces") that run functions pertaining to sorting and date parsing:

Reviewer: ```suggestion
Zoom back out to see that about 87% of CPU usage was within the `replyJSON()` method. Below that, the graph shows `replyJSON()` and the methods it calls eventually branch into four main code paths ("stack traces") that run functions pertaining to sorting and date parsing:
```

 100%    315 (longest request)
```

p99 went from 795ms to 218ms, and overall, this is four to six times faster than before.

> Locate the [profile](#read-the-profile) containing thw new load test and look at the CPU profile. The `replyJSON` parts of the flame graph are a much smaller percentage of the total CPU usage than the previous load test:

Reviewer: ```suggestion
Locate the [profile](#read-the-profile) containing the new load test and look at the CPU profile. The `replyJSON` parts of the flame graph are a much smaller percentage of the total CPU usage than the previous load test:
```

{{< img src="tracing/profiling/intro_to_profiling/flame_graph_optimized_replyjson.png" alt="Flame graph with the optimized replyJSON() stack traces">}}

### Clean up

When you're done exploring, clean up by running:

```shell
docker-compose down
```

## Recommendations

### Saving money

Improving CPU usage like this can translate into saving money. If this had been a real service, this small improvement might have enabled you to scale down to half the servers, potentially saving thousands of dollars a year. Not bad for about 10 minutes of work.

### Improve your service

This guide only skimmed the surface of profiling, but it should give you a sense of how to get started. **[Enable the Profiler for your services][8]**.

## Further reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: https://docs.docker.com/compose/install/

> [2]: https://app.datadoghq.com/account/settings#api

Reviewer: ```suggestion
[2]: https://app.datadoghq.com/organization-settings/api-keys
```
